gh-113538: Allow client connections to be closed

CendioOssman opened the pull request:
Give applications the option of more forcefully terminating client connections for asyncio servers. Useful when terminating a service and there is limited time to wait for clients to finish up their work.


<!-- gh-issue-number: gh-113538 -->
* Issue: gh-113538
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--114432.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

CendioOssman committed:
Don't use internals for wait_closed() tests

CendioOssman committed:
gh-113538: Allow client connections to be closed

Give applications the option of more forcefully terminating client
connections for asyncio servers. Useful when terminating a service and
there is limited time to wait for clients to finish up their work.

CendioOssman committed:
Fix method references in new documentation

CendioOssman committed:
Add attribution

gvanrossum reviewed (commented):
This looks great, but I have one reservation. The server now keeps the transports alive, through the set of clients. Maybe it would be better if that was a *weak* set, so that if somehow a transport was freed without going through the usual close routine, the server doesn't hold onto it forever? Thoughts?

CendioOssman committed:
Only keep a weak client references

We want to be able to detect if the application fails to keep track of
the transports, so we cannot keep them alive by using a hard reference.

CendioOssman committed:
Inform server of fallback transport cleanup

The application might be waiting for all transports to close, so we need
to properly inform the server that this transport is done.

gvanrossum commented on Lib/asyncio/base_events.py:
```diff
@@ -295,8 +296,10 @@ def _attach(self, transport):
+        # Note that 'transport' may already be missing from
+        # self._clients if it has been garbage collected
+        if transport in self._clients:
+            self._clients.remove(transport)
```
There's a simpler way: use `discard()` instead of `remove()`.
```suggestion
        self._clients.discard(transport)
```

gvanrossum commented on Doc/whatsnew/3.13.rst:
```diff
@@ -168,6 +168,11 @@ asyncio
   (Contributed by Pierre Ossman in :gh:`111246`.)
 
+* Add :meth:`asyncio.Server.close_clients` and
+  :meth:`asyncio.Server.abort_clients` methods which allows to more
```
```suggestion
  :meth:`asyncio.Server.abort_clients` methods which allow to more
```

gvanrossum commented on Lib/asyncio/base_events.py:
```diff
@@ -277,7 +277,8 @@ def __init__(self, loop, sockets, protocol_factory, ssl_context, backlog,
         self._loop = loop
         self._sockets = sockets
-        self._active_count = 0
+        # Weak references so abandoned transports can be detected
```
```suggestion
        # Weak references so abandoned transports can be ignored
```

CendioOssman replied:
The wording here was intentional. Weak references is to my knowledge the only way to detect abandoned objects. But it's not this code that does that detection, so I can understand the confusion. How about:

> Weak references so we don't break Transport's ability to detect abandoned transports

?

gvanrossum replied:
Ah, you're thinking from the POV of the transport, whose `__del__` must be called to "detect" (i.e., warn about) that it was abandoned. I was thinking from the POV of the loop in `close_clients()`, where we want to ignore (not encounter) transports that have been closed already.

I'll make it your choice which wording to use.

gvanrossum commented on Lib/asyncio/selector_events.py:
```diff
@@ -864,6 +864,8 @@ def __del__(self, _warn=warnings.warn):
         if self._sock is not None:
             _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
             self._sock.close()
+            if self._server is not None:
+                self._server._detach(self)
```
Interesting. I take it this was a bug in the original `__del__` code that you discovered during testing?

CendioOssman replied:
Indeed. If `wait_closed()` was waiting for a transport that was gc:d, then it would never finish.

gvanrossum replied:
Excellent fix.

gvanrossum commented on Misc/NEWS.d/next/Library/2024-01-22-15-50-58.gh-issue-113538.v2wrwg.rst:
```diff
@@ -0,0 +1,3 @@
+Add :meth:`asyncio.Server.close_clients` and
+:meth:`asyncio.Server.abort_clients` methods which allows to more forcefully
```
```suggestion
:meth:`asyncio.Server.abort_clients` methods which allow to more forcefully
```

gvanrossum commented on Lib/test/test_asyncio/test_server.py:
```diff
@@ -137,7 +141,8 @@ async def serve(*args):
 
         # active count != 0, not closed: should block
-        srv._attach()
+        addr = srv.sockets[0].getsockname()
+        (rd, wr) = await asyncio.open_connection(addr[0], addr[1])
```
Thanks, the calls to `_attach()`/`_detach()` in this test never sat right with me.

gvanrossum commented on Lib/test/test_asyncio/test_server.py:
```diff
@@ -161,22 +167,85 @@ async def serve(*args):
+
+        srv.close()
+        srv.close_clients()
+        await asyncio.sleep(0.1) # FIXME: flush call_soon()?
```
Yeah, short sleeps are a nuisance in asyncio tests. Usually they can be replaced by a small number of sleep(0) calls though -- usually 1, rarely 2 or 3. sleep(0) is special and guarantees we go through the event loop exactly once.

CendioOssman replied:
A bunch of `sleep(0)` felt even more arbitrary. :/

What's your suggestion here? Keep it as is? Or something else?

gvanrossum replied:
A bunch of sleep(0) wastes less time (the asyncio tests are already too slow), and doesn't risk the test becoming flaky due to timing out on slow platforms (a problem we've struggled with). The right number of sleep(0) call takes up no more time than needed, and lets the machinery go through its motions in a deterministic matter. So I recommend sleep(0).

gvanrossum commented on Lib/test/test_asyncio/test_server.py:
```diff
@@ -161,22 +167,85 @@ async def serve(*args):
+            while s_wr.transport.get_write_buffer_size() == 0:
+                s_wr.write(b'a' * 65536)
+                await asyncio.sleep(0)
+            await asyncio.sleep(0.1) # FIXME: More socket buffer space magically appears?
```
Eh, what's going on with this one? Is it the same issue that can be fixed with a small number of sleep(0) calls, or different?

CendioOssman replied:
I don't know, to be honest. Might be a kernel or libc issue where it shuffles buffers around and/or allocates more space.

Without it, I cannot reliably get both the server and client to a state where buffers are full. Which is needed for the test to check the right thing.

gvanrossum replied:
IMO this PR is not finished until you get to the bottom of that. If you really need then to reach a specific state and there's no deterministic way to get there, consider manipulating internal APIs.

CendioOssman replied:
This was a pain, but I think I got it fixed. The core problem is that the kernel dynamically increases the socket send buffer size. And it takes about 20-30 ms to do so.

I think I've worked around that by specifying an explicit buffer size. That should turn off the dynamic resizing, if I remember things correctly.

gvanrossum commented on Doc/library/asyncio-eventloop.rst:
```diff
@@ -1641,6 +1641,25 @@ Do not instantiate the :class:`Server` class directly.
 
+   .. method:: close_clients()
+
+      Close all existing incoming client connections.
```
Having now read the code, I wonder if this deserves an example or a reminder that the proper protocol is to call `close()` first, `close_clients()` second? Because if you call the latter first, there's a race condition where new connections could be made before the socket is closed.

CendioOssman replied:
Indeed. So that might be worth a note.

CendioOssman replied:
On the other hand, `close()` is synchronous when it comes to closing the incoming socket. So even if a new connection can be queueed, nothing will actually be able to call `accept()` on it. Right?

gvanrossum replied:
Agreed. It would only matter if the user called `close_clients()`, then awaited something, then called `close()`. If we reverse the order, we will know it won't matter.

gvanrossum commented on Lib/test/test_asyncio/test_server.py:
```diff
@@ -161,22 +167,85 @@ async def serve(*args):
+
+        srv.close()
+        srv.abort_clients()
+        await asyncio.sleep(0.1) # FIXME: flush call_soon()?
```
This one looks similar to the first sleep.

gvanrossum reviewed (commented):
Not bad, some thoughts.

CendioOssman committed:
Clarify recomended order of methods

CendioOssman committed:
Fix typos

CendioOssman committed:
Use discard() instead of conditional remove()

gvanrossum commented on Lib/asyncio/base_events.py:
```diff
@@ -298,8 +298,7 @@ def _attach(self, transport):
     def _detach(self, transport):
         # Note that 'transport' may already be missing from
         # self._clients if it has been garbage collected
```
IMO you don't need this comment either (it was explaining the condition).

CendioOssman committed:
Improve some comments

One could be made clearar, and the other is probably superfluous.

CendioOssman committed:
Do multiple sleeps to flush out callbacks

CendioOssman committed:
More deterministic abort_clients() test

Try to get the streams and the kernel in to a more deterministic state
by specifying fixed buffering limits.

gvanrossum commented on Lib/test/test_asyncio/test_server.py:
```diff
@@ -223,28 +224,35 @@ async def serve(rd, wr):
+        # kernel stops accepting more in to the send buffer
+        while s_wr.transport.get_write_buffer_size() == 0:
+            s_wr.write(b'a' * 4096)
+            await asyncio.sleep(0)
```
Can we put an upper bound on these `while` loops and fail the test when they loop too many times? (Possibly a `for` loop with a `break` is the best idiom to do this.)

I worry that something in the network stack might cause one or the other to loop forever, and I'd rather not waste the CPU time in CI over this.

How many iterations do you expect? Is it deterministic regardless of platform?

CendioOssman replied:
It's a bit sensitive to platform behaviour, but I believe I can get something that dynamically adapts

gvanrossum reviewed (commented):
Thanks for fixing all that. I have one remaining concern only.

CendioOssman committed:
Even more deterministic

No possibly infinite loop. Instead ask the system how much buffer space
it has and fill that.

gvanrossum reviewed (commented):
LG, but can you fix the merge conflict? The tests won't run until that's fixed.

CendioOssman committed:
Merge branch 'main' of https://github.com/python/cpython into server_close

gvanrossum reviewed (approved):
Thanks! LGTM. I'll merge.
